### PR TITLE
Add test for attributes

### PR DESCRIPTION
**Pull request by undefined-panda**
Skip write.C for versions that don't support attributes (everything before 6.40) and read.C if binary for attributes doesn't exist.

Closes #69

### CODE REVIEW

File: structure/attributes/write.C [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#include <ROOT/RNTupleModel.hxx>
+#include <ROOT/RNTupleWriteOptions.hxx>
+#include <ROOT/RNTupleWriter.hxx>
+#include <TFile.h>
```
Comment (jblomer) [inline structure/attributes/write.C]: Put in include block after the `<ROOT/...>` includes

File: structure/attributes/read.C [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#include <ROOT/RNTupleReader.hxx>
+#include <iostream>
```
Comment (jblomer) [inline structure/attributes/read.C]: Move to include block after the ROOT includes

**Review by jblomer — COMMENTED** [review]
In principle looks good to me. Since we are at it, I think we should make the test slightly more interesting and write three entries, with one attribute attached to the first entry and another attribute attached to the second and third entry.

And then, for a corner case, also test an attribute for an empty RNTuple.

File: structure/attributes/read.C [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+#include <ROOT/RVersion.hxx>
+using namespace std;
```
Comment (hahnjo) [inline structure/attributes/read.C]: Please don't put `using namespace std` into new code

File: structure/attributes/read.C [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+void read(std::string_view input = "structure.attributes.root",
+          std::string_view output = "structure.attributes.json") {
+  #if ROOT_VERSION_CODE < ROOT_VERSION(6, 40, 0)
+    cout << "Skipped structure/attributes/read.C - ROOT version too old." << endl;
+    return void();
```
Comment (hahnjo) [inline structure/attributes/read.C]: just

```suggestion
    return;
```
(or actually nothing at all, since the rest of the code is behind preprocessor guards)

File: structure/attributes/read.C [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      // next entry.
+    }
+
+    auto attrSet = reader->OpenAttributeSet("Attributes");
+    auto attrEntry = attrSet->CreateEntry();
+    auto attrInt32 = attrEntry->GetPtr<std::int32_t>("attrInt32");
+    
+    for (auto idx : reader->GetEntryRange()) {
+      for (auto attrIdx : attrSet->GetAttributes()) {
+        attrSet->LoadEntry(attrIdx, *attrEntry);
```
Comment (hahnjo) [inline structure/attributes/read.C]: This prints all attributes for all entries in the main RNTuple. That's probably not what we want, also in light of Jakob's comment to have multiple attribute entries. We also need to print the attribute range(s)

File: structure/attributes/read.C [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    } catch (const ROOT::RException &e) {
+      std::ofstream os(std::string{output});
+      std::string msgWithoutStacktrace;
+      std::getline(std::istringstream(e.what()), msgWithoutStacktrace);
+      os << "{\n";
+      os << "  \"error\": \"" << msgWithoutStacktrace << "\"\n";
+      os << "}\n";
+      return;
+    } catch (const runtime_error &e) {
+      cout << "Skipped structure/attributes/read.C - " << e.what() << endl;
```
Comment (hahnjo) [inline structure/attributes/read.C]: Should we handle the case of missing `structure.attributes.root` more explicitly? That is the only error that I think is permissable, all others should make the test fail.

File: structure/attributes/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# Attributes
+
+A RNTuple with a regular field and a attribute field.
```
Comment (hahnjo) [inline structure/attributes/README.md]: Please describe at least the attribute name and its field in more detail, similar to the explanations for other tests.

File: structure/attributes/read.C [hunk 8]
```diff
@@ -0,0 +1,70 @@
+#include <ROOT/RNTupleReader.hxx>
+#include <iostream>
+#include <ROOT/RVersion.hxx>
+using namespace std;
+
+void read(std::string_view input = "structure.attributes.root",
+          std::string_view output = "structure.attributes.json") {
+  #if ROOT_VERSION_CODE < ROOT_VERSION(6, 40, 0)
```
Comment (hahnjo) [inline structure/attributes/read.C]: In general, please run `clang-format` on the C++ code.

File: structure/attributes/write.C (hunk 1 above)
Comment (undefined-panda) [reply structure/attributes/write.C]: Moved include block

File: structure/attributes/read.C (hunk 2 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Moved include block

File: structure/attributes/read.C (hunk 3 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Removed `using namespace`

File: structure/attributes/read.C (hunk 4 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Removed return

File: structure/attributes/read.C [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    os << "      {\n";
+    os << "        \"Value\": " << *attr << ",\n";
+    os << "        \"Range\": [" << *range.GetFirst() << ", " << *range.GetLast() << "]\n";
+    if (first) {
+      os << "      },\n";
+      first = false;
+    } else {
+      os << "      }\n";
+    }    
+    std::cout << "  attr = " << *attr << " (valid for range [" << *range.GetFirst() << ", " << *range.GetLast() << "])\n";
```
Comment (hahnjo) [inline structure/attributes/read.C]: Is this still debugging output?

File: structure/attributes/read.C [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+#include <iostream>
+
+void read(std::string_view input = "structure.attributes.root",
+          std::string_view output = "structure.attributes.json") {
+#if ROOT_VERSION_CODE < ROOT_VERSION(6, 40, 0)
+  std::cout << "Skipped structure/attributes/read.C - ROOT version too old."
+            << std::endl;
+#else
+  // same logic as is read_structure.hxx, with additional special part for
+  // attributes
```
Comment (hahnjo) [inline structure/attributes/read.C]: Actually, the latest code doesn't print the main entry anymore. Probably we don't need this, but then we should update / remove this comment.

File: structure/attributes/read.C [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    return;
+  } catch (const runtime_error &e) {
+    cout << "Skipped structure/attributes/read.C - " << e.what() << endl;
+    return;
+  }
+
+  std::ofstream os(std::string{output});
+  os << "[\n";
+  auto Int32 = 
+      reader->GetModel().GetDefaultEntry().GetPtr<std::int32_t>("Int32");
```
Comment (hahnjo) [inline structure/attributes/read.C]: unused

File: structure/attributes/read.C [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  os << "    \"Attribute set name\": \"" << attrSet->GetDescriptor().GetName() << "\"\n";
+  os << "  }, \n";
+  os << "  {\n";
+  os << "    \"Description\": \"" << attrSet->GetDescriptor().GetDescription() << "\"\n";
+  os << "  },\n";
+
+  // list of values for attr with its range
+  os << "  {\n";
+  os << "    \"Attr\": " << "[\n";
+  first = true;
```
Comment (hahnjo) [inline structure/attributes/read.C]: already set above

File: structure/attributes/README.md [hunk 13]
```diff
@@ -0,0 +1,16 @@
+# Attributes
+
+## Fields
+
+ * `Int{32}`
```
Comment (hahnjo) [inline structure/attributes/README.md]:
```suggestion
 * `Int32`
```

File: structure/attributes/read.C (hunk 6 above)
Comment (hahnjo) [reply structure/attributes/read.C]: AFAICT the code hasn't changed, a missing `structure.attributes.root` is not handled explicitly...

File: structure/attributes/README.md [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+ * `Attr`
+
+one regular and one attribute field.
+
+## Entries
+
+1. Attribute set name
+2. Attribute set description
+3. List of attribute entries with value and regular field range
```
Comment (hahnjo) [inline structure/attributes/README.md]: These are not (RNTuple) entries. What you want to describe here is that there is one attribute entry spanning one main entry, and the second attribute entry spanning two main entries.

File: structure/attributes/README.md [hunk 15]
```diff
@@ -0,0 +1,16 @@
+# Attributes
+
+## Fields
+
+ * `Int{32}`
+ * `Attr`
```
Comment (hahnjo) [inline structure/attributes/README.md]: In the code it's "attr" (lower-case). Maybe you can also make it explicit that `Int32` is the field in the main RNTuple, and `attr` is the field in the attribute model

File: structure/attributes/read.C (hunk 6 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Explicit error handling for missing file was added.

File: structure/attributes/read.C (hunk 9 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Removed

File: structure/attributes/read.C (hunk 10 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Comment updated

File: structure/attributes/read.C (hunk 11 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Removed

File: structure/attributes/read.C (hunk 12 above)
Comment (hahnjo) [reply structure/attributes/read.C]: still already set above - please remove

File: structure/attributes/README.md (hunk 14 above)
Comment (hahnjo) [reply structure/attributes/README.md]: The first two items of the list are still there. This section of the README should describe what is written / read from the RNTuple data, ie "unroll" the third item.

File: structure/attributes/write.C [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+#else
+  auto model = ROOT::RNTupleModel::Create();
+  auto Int32 = model->MakeField<std::int32_t>("Int32");
+
+  auto file = std::unique_ptr<TFile>(TFile::Open(std::string(filename).c_str(), "RECREATE"));
+  auto writer = ROOT::RNTupleWriter::Append(std::move(model), "ntpl", *file);
+
+  // definition of attribute set with attribute fields
+  auto attrModel = ROOT::RNTupleModel::Create();
+  attrModel->SetDescription("Attribute set for testing");
```
Comment (hahnjo) [inline structure/attributes/write.C]: Actually do we want to test this? Right now we don't do it for the main RNTuple descriptor, nor for the field descriptions etc...

File: structure/attributes/read.C [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+  std::ofstream os(std::string{output});
+  os << "[\n";
+  bool first = true;
+
+  auto attrSet = reader->OpenAttributeSet("Attributes");
+  auto attr = attrSet->GetModel().GetDefaultEntry().GetPtr<std::int32_t>("attr");
+
+  os << "  {\n";
+  os << "    \"Attribute set name\": \"" << attrSet->GetDescriptor().GetName() << "\"\n";
```
Comment (hahnjo) [inline structure/attributes/read.C]: This is not really useful since the name is already hard-coded above

File: structure/attributes/read.C [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+  os << "  }, \n";
+  os << "  {\n";
+  os << "    \"Description\": \"" << attrSet->GetDescriptor().GetDescription() << "\"\n";
```
Comment (hahnjo) [inline structure/attributes/read.C]: see other comment

File: structure/attributes/read.C [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+  os << "  },\n";
+
+  // list of values for attr with its range
+  os << "  {\n";
+  os << "    \"Attr\": " << "[\n";
+  first = true;
+  for (auto attrIdx : attrSet->GetAttributes()) {
```
Comment (hahnjo) [inline structure/attributes/read.C:28]: These entries should be the main content of the `.json` file

File: structure/attributes/README.md [hunk 20]
```diff
@@ -0,0 +1,16 @@
+# Attributes
+
+## Fields
+
+ * `Int32` (in main RNTuple)
+ * `attr` (in attribute model)
+
+one regular and one attribute field.
```
Comment (hahnjo) [inline structure/attributes/README.md]: This information is now in the list. However, we should describe the field types (`std::int32_t`)

File: structure/attributes/read.C [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    reader = ROOT::RNTupleReader::Open("ntpl", input);
+  } catch (const ROOT::RException &e) {
+    std::ofstream os(std::string{output});
+    std::string msgWithoutStacktrace;
+    std::getline(std::istringstream(e.what()), msgWithoutStacktrace);
+    os << "{\n";
+    os << "  \"error\": \"" << msgWithoutStacktrace << "\"\n";
+    os << "}\n";
+    return;
+  }
```
Comment (hahnjo) [inline structure/attributes/read.C]: Actually I think we can simplify this: in `read_structure.hxx`, we catch exceptions because `structure/feature_flag/read.C` requires it. Here we can go back to the naive code of just calling `RNTupleReader::Open` and in case of exceptions let the entire macro fail.

File: structure/attributes/read.C (hunk 12 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Removed

File: structure/attributes/read.C (hunk 18 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Removed

File: structure/attributes/read.C (hunk 17 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Removed

File: structure/attributes/read.C (hunk 21 above)
Comment (undefined-panda) [reply structure/attributes/read.C]: Removed try-catch block, s.t. just `RNTupleReader::Open` is being called.

File: structure/attributes/write.C (hunk 16 above)
Comment (undefined-panda) [reply structure/attributes/write.C]: Removed description

File: structure/attributes/read.C [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+            << std::endl;
+#else
+  // skipping test if .root file file doesn't exist
+  if (!std::filesystem::exists(input)) {
+    std::cout << "Skipped structure/attributes/read.C - file ('" << input << "') not found" << std::endl;
+    return;
+  }
+  // similar logic as read_structure.hxx
+  std::unique_ptr<ROOT::RNTupleReader> reader;
+  reader = ROOT::RNTupleReader::Open("ntpl", input);
```
Comment (hahnjo) [inline structure/attributes/read.C]:
```suggestion
  std::unique_ptr<ROOT::RNTupleReader> reader = ROOT::RNTupleReader::Open("ntpl", input);
```

File: structure/attributes/read.C [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+  std::ofstream os(std::string{output});
+  os << "[\n";
+  bool first = true;
+
+  auto attrSet = reader->OpenAttributeSet("Attributes");
+  auto attr = attrSet->GetModel().GetDefaultEntry().GetPtr<std::int32_t>("attr");
+
+  // list of values for attr with its range
+  os << "  {\n";
+  os << "    \"Attr\": " << "[\n";
```
Comment (hahnjo) [inline structure/attributes/read.C]: You don't need a subobject anymore, just write the attributes into the top-level array. I think the JSON should look like this:
```json
[
  {
    "Value": 1,
    "Range": [0, 0]
  },
  {
    "Value": 2,
    "Range": [1, 2]
  }
}
```

File: structure/attributes/README.md [hunk 24]
```diff
@@ -0,0 +1,14 @@
+# Attributes
+
+## Fields
+
+ * `Int32` (in main RNTuple)
+ * `attr[std::int32_t]` (in attribute model)
+
+one regular and one attribute field.
```
Comment (hahnjo) [inline structure/attributes/README.md]:
```suggestion
 * `Int32` (in main RNTuple)
 * `attr` (in attribute set called `Attributes`)

with each the type `std::int32_t`.
```

File: structure/attributes/README.md [hunk 25]
```diff
@@ -0,0 +1,14 @@
...
+
+## Entries
+
+1. Values for `attr`. One attribute entry refers to first main entry, a second one to next two main entries
```
Comment (hahnjo) [inline structure/attributes/README.md]:
```suggestion
## Attribute Entries

1. Associated to the first entry of the main RNTuple
2. Referring to the second and third entry of the main RNTuple
```

File: Makefile [hunk 26]
```diff
@@ -71,9 +71,10 @@ validate:
 	
 #	run make read to do cross-validation
 	@for s_path in $(source_scripts); do \
-		for w_dir in write/*; do \
-			(bash -c "source $$s_path && ROOT_VERSION=\$$(root-config --version) && $(MAKE) read dict_dir=\$$ROOT_VERSION write_dir=$$(basename $$w_dir) read_dir=\$$ROOT_VERSION") || exit $$?; \
-		done; \
+		bash -c "source $$s_path && ROOT_VERSION=\$$(root-config --version) && \
+			for w_dir in write/*; do \
+				$(MAKE) read dict_dir=\$$ROOT_VERSION write_dir=\$$(basename \$$w_dir) read_dir=\$$ROOT_VERSION || exit \$$?; \
+			done" || exit $$?; \
```
Comment (hahnjo) [inline Makefile]: This should go into a separate PR

File: structure/attributes/read.C [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+    auto range = attrSet->LoadEntry(attrIdx);
+    os << "  {\n";
+    os << "    \"Value\": " << *attr << ",\n";
+    os << "    \"Range\": [" << *range.GetFirst() << ", " << *range.GetLast() << "]\n";
+    if (first) {
+      os << "   },\n";
+      first = false;
+    } else {
+      os << "   }\n";
+    }
```
Comment (hahnjo) [inline structure/attributes/read.C:34]: Hm wait, this only produces valid JSON for exactly two attribute entries. Can you please apply the approach taken in other tests (such as `types/fundamental/integer/read.C`), to print the comma only in the next iteration?

File: structure/attributes/README.md [hunk 28]
```diff
@@ -0,0 +1,15 @@
+# Attributes
+
+## Fields
+
+ * `Int32` (in main RNTuple)
+ * `attr[std::int32_t]` (in attribute set called `Attributes`)
```
Comment (hahnjo) [inline structure/attributes/README.md]:
```suggestion
 * `attr` (in attribute set called `Attributes`)
```

**Review by hahnjo — COMMENTED** [review]
Sorry, last two things I just noticed

**Review by hahnjo — APPROVED** [review]
